File: problem-solving/_lc-0198-medium-house-robber/1-top-down.go
/*
https://leetcode.com/problems/house-robber/

Complexity
----------
Time: O(n)
Space: O(n)
*/

func rob(nums []int) int {
    cache := make(map[int]int)
    var robMax func(int) int
    robMax = func(i int) int {
        if i >= len(nums) {
            return 0
        }
        if v, ok := cache[i]; ok {
            return v
        }
        cache[i] = max(
            nums[i] + robMax(i + 2),
            robMax(i + 1))
        return cache[i]
    }
    return robMax(0)
}

func max(x, y int) int {
    if x > y {
        return x
    }
    return y
}

File: problem-solving/_lc-198-medium-house-robber/1-top-down.py
"""
https://leetcode.com/problems/house-robber/

Complexity
----------
Time: O(n)
Space: O(n)
"""

class Solution:
    def rob(self, nums: List[int]) -> int:
        cache = {}
        def robMax(idx):
            if idx >= len(nums):
                return 0
            if idx in cache:
                return cache[idx]
            notTaken = robMax(idx + 1)  # Skip current but rob next
            taken = nums[idx] + robMax(idx + 2)  # Rob current and skip next
            cache[idx] = max(taken, notTaken)
            return cache[idx]
        return robMax(0)
